src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php

```diff
@@ -82,8 +114,13 @@ public function getTypeGuesser()
             $this->guesserLoaded = true;
             $guessers = array();
 
-            foreach ($this->guesserServiceIds as $serviceId) {
-                $guessers[] = $this->container->get($serviceId);
+            foreach ($this->guesserServices as $serviceId => $service) {
+                if (is_string($service)) {
+                    $serviceId = $service;
+                    $service = $this->typeContainer->get($serviceId);
+                }
+
+                $guessers[] = $service;
```

Comment:
> It's just a bit more difficult to read and understand the code at the moment because of the mixed two layers.

Indeed :)  
I'm unable to say if the BC layer is correct right now, but the approach looks better!